File: src/main/java/kr/co/bookand/backend/policy/service/PolicyService.java
package kr.co.bookand.backend.policy.service;

import kr.co.bookand.backend.account.domain.Account;
import kr.co.bookand.backend.account.service.AccountService;
import kr.co.bookand.backend.common.exception.ErrorCode;
import kr.co.bookand.backend.policy.domain.Policy;
import kr.co.bookand.backend.policy.domain.dto.PolicyDto;
import kr.co.bookand.backend.policy.exception.PolicyException;
import kr.co.bookand.backend.policy.repository.PolicyRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;

import static kr.co.bookand.backend.policy.domain.dto.PolicyDto.*;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class PolicyService {

    private final PolicyRepository policyRepository;
    private final AccountService accountService;

    @Transactional
    public PolicyResponse updatePolicy(Long id, PolicyRequest request) {
        Account currentAccount = accountService.getCurrentAccount();
        currentAccount.getRole().checkAdmin();
        Policy policy = policyRepository.findById(id).orElseThrow(() -> new PolicyException(ErrorCode.NOT_FOUND_POLICY, request));
        policy.updateContent(request.content());
        return PolicyResponse.of(policy);
    }

    public PolicyResponse getTitlePolicy(String name) {
        return policyRepository.findByName(name).map(PolicyResponse::of).orElseThrow(() -> new PolicyException(ErrorCode.NOT_FOUND_POLICY, name));
    }

    public PolicyResponse getPolicy(Long id) {
        return policyRepository.findById(id).map(PolicyResponse::of).orElseThrow(() -> new PolicyException(ErrorCode.NOT_FOUND_POLICY, id));
    }

    @Transactional
    public PolicyResponse createPolicy(PolicyRequest request) {
        Account currentAccount = accountService.getCurrentAccount();
        currentAccount.getRole().checkAdmin();
        Optional<Policy> policyResponse = policyRepository.findByName(request.name());
        if (policyResponse.isPresent()) {
            throw new PolicyException(ErrorCode.ALREADY_EXIST_POLICY, request);
        }
        Policy policy = request.toPolicy();
        Policy save = policyRepository.save(policy);
        return PolicyResponse.of(save);
    }

    public void removePolicy(Long id) {
        Account currentAccount = accountService.getCurrentAccount();
        currentAccount.getRole().checkAdmin();
        Policy policy = policyRepository.findById(id).orElseThrow(() -> new PolicyException(ErrorCode.NOT_FOUND_POLICY, id));
        policyRepository.delete(policy);
    }
}

File: src/main/kotlin/kr/co/bookand/backend/policy/service/PolicyService.kt
package kr.co.bookand.backend.policy.service


import kr.co.bookand.backend.account.domain.Account
import kr.co.bookand.backend.common.ErrorCode
import kr.co.bookand.backend.common.exception.BookandException
import kr.co.bookand.backend.policy.domain.Policy
import kr.co.bookand.backend.policy.domain.dto.PolicyIdResponse
import kr.co.bookand.backend.policy.domain.dto.CreatePolicyRequest
import kr.co.bookand.backend.policy.domain.dto.PolicyResponse
import kr.co.bookand.backend.policy.repository.PolicyRepository
import lombok.RequiredArgsConstructor
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
class PolicyService(
    private val policyRepository: PolicyRepository
) {
    @Transactional
    fun createPolicy(currentAccount: Account, request: CreatePolicyRequest): PolicyIdResponse {
        currentAccount.role.checkAdminAndManager()
        val policy = getPolicyByName(request.name)
        if (policy != null) {
            throw BookandException(ErrorCode.ALREADY_EXIST_POLICY)
        }
        val newPolicy = Policy(
            title = request.title,
            name = request.name,
            content = request.content
        )
        val savePolicy = policyRepository.save(newPolicy)
        return PolicyIdResponse(savePolicy.id)
    }

    @Transactional
    fun updatePolicy(
        currentAccount: Account,
        id: Long,
        request: CreatePolicyRequest
    ): PolicyIdResponse {
        currentAccount.role.checkAdminAndManager()
        val policy = getPolicyByIdOrThrow(id)
        policy.updateContent(request.content)
        return PolicyIdResponse(policy.id)
    }

    fun getTitlePolicy(name: String): PolicyResponse {
        val policy = getPolicyByNameOrThrow(name)
        return PolicyResponse(policy)
    }

    fun removePolicy(currentAccount: Account, id: Long) {
        currentAccount.role.checkAdminAndManager()
        val policy = getPolicyByIdOrThrow(id)
        policyRepository.delete(policy)
    }

    private fun getPolicyByIdOrThrow(id: Long): Policy {
        return policyRepository.findById(id).orElse(null)
            ?: throw BookandException(ErrorCode.NOT_FOUND_POLICY)
    }

    private fun getPolicyByNameOrThrow(name: String): Policy {
        return policyRepository.findByName(name)
            ?: throw BookandException(ErrorCode.NOT_FOUND_POLICY)
    }

    private fun getPolicyByName(name: String): Policy? {
        return policyRepository.findByName(name)
    }
}